diff --git a/gcc/config.build b/gcc/config.build
index 7451226f6ac..098b48f0158 100644
--- a/gcc/config.build
+++ b/gcc/config.build
@@ -53,7 +53,7 @@ case $build in
   hppa*64*-*-hpux11* )
     build_install_headers_dir=install-headers-cpio
     ;;
-  i[34567]86-*-cygwin* | x86_64-*-cygwin* | aarch64-*-cygwin*)
+  i[34567]86-*-cygwin* | x86_64-*-cygwin* | aarch64-*-cygwin* | aarch64-*-msys*)
     build_xm_file=i386/xm-cygwin.h
     build_exeext=.exe
     ;;
diff --git a/gcc/config.gcc b/gcc/config.gcc
index 008a7606782..13f6cab094d 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -1275,6 +1275,36 @@ aarch64*-*-gnu*)
         tmake_file="${tmake_file} aarch64/t-aarch64"
         tm_defines="${tm_defines}  TARGET_DEFAULT_ASYNC_UNWIND_TABLES=1"
 	;;
+aarch64-*-msys*)
+	tm_file="${tm_file} aarch64/aarch64-abi-ms.h"
+	tm_file="${tm_file} aarch64/aarch64-coff.h"
+	tm_file="${tm_file} aarch64/cygming.h"
+	tm_file="${tm_file} i386/cygwin.h"
+	tm_file="${tm_file} i386/cygwin-w64.h"
+	tm_file="${tm_file} i386/cygwin-stdint.h"
+	tm_file="${tm_file} mingw/winnt.h"
+	tm_file="${tm_file} mingw/winnt-dll.h"
+	xm_file=i386/xm-cygwin.h
+	tmake_file="${tmake_file} aarch64/t-aarch64"
+	tmake_file="${tmake_file} mingw/t-cygming t-slibgcc"
+	target_gtfiles="$target_gtfiles \$(srcdir)/config/mingw/winnt.cc"
+	target_gtfiles="$target_gtfiles \$(srcdir)/config/mingw/winnt-dll.cc"
+	extra_options="${extra_options} mingw/cygming.opt i386/cygwin.opt"
+	extra_objs="${extra_objs} winnt.o winnt-dll.o"
+	c_target_objs="${c_target_objs} msformat-c.o"
+	cxx_target_objs="${cxx_target_objs} winnt-cxx.o msformat-c.o"
+	d_target_objs="${d_target_objs} cygwin-d.o"
+	target_has_targetdm="yes"
+	aarch64_multilibs="llp64"
+	if test x$enable_threads = xyes; then
+		thread_file='posix'
+	fi
+	default_use_cxa_atexit=yes
+	use_gcc_stdint=wrap
+	tm_defines="${tm_defines} TARGET_64BIT=1 TARGET_MSYS=1 TARGET_64BIT_MS_ABI=1 TARGET_AARCH64_MS_ABI=1"
+	TM_MULTILIB_CONFIG="${TM_MULTILIB_CONFIG},llp64"
+	TM_MULTILIB_CONFIG=`echo $TM_MULTILIB_CONFIG | sed 's/^,//'`
+	;;
 aarch64-*-cygwin*)
 	tm_file="${tm_file} aarch64/aarch64-abi-ms.h"
 	tm_file="${tm_file} aarch64/aarch64-coff.h"
diff --git a/gcc/config.host b/gcc/config.host
index 3fe37fe4b57..2adc44dedb0 100644
--- a/gcc/config.host
+++ b/gcc/config.host
@@ -228,7 +228,7 @@ case ${host} in
         ;;
     esac
     ;;
-  i[34567]86-*-cygwin* | x86_64-*-cygwin* | aarch64-*-cygwin*)
+  i[34567]86-*-cygwin* | x86_64-*-cygwin* | aarch64-*-cygwin* | aarch64-*-msys* )
     host_xm_file=i386/xm-cygwin.h
     out_host_hook_obj=host-cygwin.o
     host_xmake_file="${host_xmake_file} i386/x-cygwin"
diff --git a/libgcc/config.host b/libgcc/config.host
index 0ae68b669f1..3981ae22dc7 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -866,6 +866,28 @@ i[34567]86-*-solaris2* | x86_64-*-solaris2*)
 	;;
 i[4567]86-wrs-vxworks*|x86_64-wrs-vxworks*)
 	;;
+aarch64-*-msys*)
+	if test x$ac_cv_sjlj_exceptions = xyes; then
+		tmake_eh_file="i386/t-sjlj-eh"
+	else
+		tmake_eh_file="mingw/t-seh-eh"
+	fi
+	if test x${build} = x${host} ; then
+		tmake_dlldir_file="i386/t-dlldir"
+	else
+		tmake_dlldir_file="i386/t-dlldir-x"
+	fi
+	tmake_file="${tmake_file} ${tmake_eh_file} ${tmake_dlldir_file}"
+	tmake_file="${tmake_file} t-dfprules"
+	tmake_file="${tmake_file} i386/t-slibgcc-cygming i386/t-cygming i386/t-cygwin"
+	tmake_file="${tmake_file} ${cpu_type}/t-aarch64"
+	tmake_file="${tmake_file} ${cpu_type}/t-lse"
+	tmake_file="${tmake_file} ${cpu_type}/t-softfp t-softfp t-crtfm"
+	extra_parts="$extra_parts crtbegin.o crtbeginS.o crtend.o crtfastmath.o"
+	if test x$enable_vtable_verify = xyes; then
+		extra_parts="$extra_parts vtv_start.o vtv_end.o vtv_start_preinit.o vtv_end_preinit.o"
+	fi
+	;;
 aarch64-*-cygwin*)
 	# This has to match the logic for DWARF2_UNWIND_INFO in gcc/config/aarch64/cygming.h
 	if test x$ac_cv_sjlj_exceptions = xyes; then
